common/functions-v2/src/scalars/string.rs

```diff
@@ -77,22 +98,31 @@ pub fn register(registry: &mut FunctionRegistry) {
     registry.register_aliases("char_length", &["character_length"]);
 }
 
+// Vectorize string to string function with default estimate_bytes, which is the original string length.
 fn vectorize_string_to_string(
-    estimate_bytes: impl Fn(&StringColumn) -> usize + Copy,
-    func: impl Fn(&[u8], &mut StringColumnBuilder) -> Result<(), String> + Copy,
+    func: impl Fn(&[u8], &mut [u8]) -> Result<usize, String> + Copy,
+) -> impl Fn(ValueRef<StringType>, &GenericMap) -> Result<Value<StringType>, String> + Copy {
+    vectorize_string_to_string_wise(|(data, _)| data.len(), func)
+}
+
+// Vectorize string to string function with customer estimate_bytes.
+fn vectorize_string_to_string_wise(
+    estimate_bytes: impl Fn((&[u8], &Buffer<u64>)) -> usize + Copy,
+    func: impl Fn(&[u8], &mut [u8]) -> Result<usize, String> + Copy,
```

Comment:
Would it be better if we use `StringColumn` here?

Reply:
I want the `Builder` of `Scalar` also create with `estimate_bytes`, so I treat `scalar` as `a one item column`.

If we use `&StringColumn` as the arg of `estimate_bytes`, it has to clone the data to transfer `&[u8]` to `StringColumn`.

So I use `(&[u8], &Buffer<u64>)` to avoid data clone, which is equivalent to `&StringColumn`.

Maybe we can define it as a struct like  `StringColumnRef` for readability.

Reply:
In fact, `Buffer<T>` is just a smart point to the underlying data, which means the clone is nearly zero-cost.

Because functions for Scalar are usually used in constant folding, I think it doesn't worth optimizing the allocation for Scalars.

Reply:
Let's remove scalar optimize and use `StringColumn` here.

Reply:
Ohh, I found the core issue here.

Since both scalar and column branches use the same func in `vectorize`, we also need to get mutable slice in `scalar` branch, then we are forced to `estimate_bytes`.

Reply:
We could add a method to `StringColumnBuilder` like:


```rust
pub fn alloc_slice(&mut self, len: usize) -> &mut [u8] { ... }
```

Reply:
I suggest using [base64-stream]( https://github.com/magiclen/base64-stream) instead

Reply:
What is the difference between these two?

Reply:
`base64-stream` can output encode result to `&mut Vec<u8>`

Reply:
Just realized thaT `base64-stream` doesn't encode with padding...

Reply:
> Just realized that base64-stream doesn't encode with padding...

Do you mean we can change the write to `&mut Vec<u8>`, then `std::slice::from_raw_parts_mut` is no needed any more?

```rust
fn vectorize_string_to_string(
    estimate_bytes: impl Fn((&[u8], Buffer<u64>)) -> usize + Copy,
    func: impl Fn(&[u8], &mut Vec<u8>) -> Result<usize, String> + Copy,
) -> impl Fn(ValueRef<StringType>, &GenericMap) -> Result<Value<StringType>, String> + Copy {
```

Reply:
Seems it's more reasonable.

Reply:
I [took a try at it](https://github.com/ygf11/databend/pull/1), and it seems to work.